[1.10] Can't push to private registry

Running a registry (`registry:2.3`) on `somehost:5000` with TLS. 

Pushing the images from docker 1.9
```
docker --version
Docker version 1.9.1, build a34a1d5
``` 
works as expected.

When pushing from docker 1.10 
```
docker --version
Docker version 1.10.0, build 590d5108
```

```
docker push somehost:5000/someimage
The push refers to a repository [somehost:5000/someimage]
88845c8d6bba: Pushed 
ee2754fa8a70: Pushed 
af802994b36b: Pushed 
d570987b7ef3: Pushed 
d889f924f1e7: Pushed 
f1f9da510ceb: Pushed 
12bfb8bedc56: Pushed 
d9a5758a1233: Pushed 
553f5b0fe85e: Pushed 
bacf371c2eee: Pushed 
c8e22318f4c4: Pushed 
01ec7f26f836: Pushed 
095dc581fddf: Pushed 
5f70bf18a086: Pushed 
314fc1ee156b: Pushed 
4d6aa1aa583e: Pushed 
cde9196e5cc7: Pushed 
283b24c4a7bb: Pushed 
f5c1535f7dcf: Pushed 
6859151bd499: Pushed 
2c63ca40402a: Pushed 
c48b7334bb1a: Pushed 
1f253a0703ec: Pushing 1.895 kB/1.895 kB
80ec921b230b: Retrying in 4 seconds 
227021bc9aa6: Pushing 32.66 MB/187.7 MB
Get http://somehost:5000/v2/: malformed HTTP response "\x15\x03\x01\x00\x02\x02"

I'm getting the same problem.   


`docker version`:
Docker version 1.10.1, build 9e83765

`docker info`:
Containers: 3
 Running: 1
 Paused: 0
 Stopped: 2
Images: 1834
Server Version: 1.10.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1852
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.16.0-41-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.734 GiB
Name: xen
ID: IOQK:UNKL:TKWG:GC76:TISS:HPYN:XECB:7VEV:RWBE:WCWI:ACAI:KTRR
WARNING: No swap limit support

`uname -a`:
3.16.0-41-generic

```
matthewh@xen:~/dev/cluster/docker/ibdiags$ docker build -t localhost:5000/matthewh/ibdiags:latest .
Sending build context to Docker daemon 737.8 kB
Step 1 : FROM ubuntu:15.04
 ---> 450463fde31f
Step 2 : RUN apt-get -qq update &&     apt-get -qq install infiniband-diags ibutils &&     apt-get clean && apt-get autoremove
 ---> Using cache
 ---> 5ff9f48fdbcf
Step 3 : CMD /bin/bash
 ---> Using cache
 ---> 68021b396974
Successfully built 68021b396974

matthewh@xen:~/dev/cluster/docker/ibdiags$ docker push localhost:5000/matthewh/ibdiags:latest 
The push refers to a repository [localhost:5000/matthewh/ibdiags]
5f70bf18a086: Pushing 
b707364e164c: Pushing 113.3 MB/115.9 MB
ed58a6b8d8d6: Pushing 1.879 kB/1.879 kB
84cc3d400b0d: Pushing    701 B/701 B
3cbe18655eb6: Pushing 17.38 MB/131.3 MB
unexpected EOF
make: *** [all] Error 1
```

Also failed when trying to push to Amazon ECS.

Next step: purge and remove all images with docker stop docker rm and docker rmi

Verify nothing left:

``` 
matthewh@xen:~/dev/cluster/docker$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
matthewh@xen:~/dev/cluster/docker$ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
matthewh@xen:~/dev/cluster/docker$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
```

Start a new registry for testing:

```
matthewh@xen:~$ docker run -d -p 5000:5000 --restart=always --name registry registry:2
Unable to find image 'registry:2' locally
2: Pulling from library/registry
7268d8f794c4: Pull complete 
a3ed95caeb02: Pull complete 
6915c066a24f: Pull complete 
7f04289ca403: Pull complete 
7539a42f6147: Pull complete 
Digest: sha256:339d702cf9a4b0aa665269cc36255ee7ce424412d56bee9ad8a247afe8c49ef1
Status: Downloaded newer image for registry:2
3a5c1108c032c175c41a760b21e6a27f7ccbf15a521301239d2ef1883aebc504
```

Now rebuild earlier from scratch and see if we can push.
```
docker build -t localhost:5000/matthewh/ibdiags:latest .
Sending build context to Docker daemon 737.8 kB
Step 1 : FROM ubuntu:15.04
15.04: Pulling from library/ubuntu
9502adfba7f1: Pull complete 
4332ffb06e4b: Pull complete 
2f937cc07b5f: Pull complete 
a3ed95caeb02: Pull complete 
Digest: sha256:2fb27e433b3ecccea2a14e794875b086711f5d49953ef173d8a03e8707f1510f
Status: Downloaded newer image for ubuntu:15.04
 ---> d1b55fd07600
Step 2 : RUN apt-get -qq update &&     apt-get -qq install infiniband-diags ibutils &&     apt-get clean && apt-get autoremove
 ---> Running in 8bfecbdf50f6

--- snip ---

0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
 ---> 7a3709808c06
Removing intermediate container 8bfecbdf50f6
Step 3 : CMD /bin/bash
 ---> Running in e625e16bbb96
 ---> 0e60465fdeea
Removing intermediate container e625e16bbb96
Successfully built 0e60465fdeea

matthewh@xen:~/dev/cluster/docker$ docker push localhost:5000/matthewh/ibdiags:latest 
The push refers to a repository [localhost:5000/matthewh/ibdiags]
a9b6fb6e858b: Pushed 
5f70bf18a086: Pushed 
ed58a6b8d8d6: Pushed 
84cc3d400b0d: Pushed 
3cbe18655eb6: Pushed 
latest: digest: sha256:90e3458bb5b2ff1c0b4cc7c443416d8627d47d1c4cb5bb3e4c061c14ed1b0c18 size: 1339
```

Purging the local registry seems to be the workaround.

--- a/layer/migration.go
+++ b/layer/migration.go
@@ -127,6 +127,7 @@ func (ls *layerStore) checksumForGraphIDNoTarsplit(id, parent, newTarDataPath st
 	}
 	defer f.Close()
 	mfz := gzip.NewWriter(f)
+	defer mfz.Close()
 	metaPacker := storage.NewJSONPacker(mfz)
 
 	packerCounter := &packSizeCounter{metaPacker, &size}

